### PR TITLE
rework containerd config merging

### PR DESCRIPTION
**Pull request by ndbaker1**
**Issue #, if available:**

**Description of changes:**

It was recently realized that containerd's native import merging strategy is not one that behaves as most users would like. This is because it replaces the entire body under keyed maps such as `Plugins`. This conflicts with the use case to override singular parameters under existing sections. (see [xref](https://github.com/containerd/containerd/blob/main/cmd/containerd/server/config/config.go#L407-L431)). 

This PR implements the merging in an unstructured way so that users can continue providing whatever inline containerd config fields they would like overridden.

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

**Testing Done**

*[See this guide for recommended testing for PRs.](../doc/CONTRIBUTING.md#testing-changes) Some tests may not apply. Completing tests and providing additional validation steps are not required, but it is recommended and may reduce review time and time to merge.*

### CODE REVIEW

**Comment by ndbaker1** [comment]
/ci
+workflow:os_distros al2023

File: nodeadm/internal/containerd/config.go [hunk 1]
```diff
@@ -35,16 +36,28 @@ func writeContainerdConfig(cfg *api.NodeConfig) error {
 	if err != nil {
 		return err
 	}
-	zap.L().Info("Writing containerd config to file..", zap.String("path", containerdConfigFile))
-	if err := util.WriteFileWithDir(containerdConfigFile, containerdConfig, containerdConfigPerm); err != nil {
-		return err
-	}
+	// because the logic in containerd's import merge decides to completely
+	// overwrite entire sections, we want to implement this merging ourselves.
+	// see: https://github.com/containerd/containerd/blob/main/cmd/containerd/server/config/config.go#L407-L431
```
Comment (cartermckinnon) [inline nodeadm/internal/containerd/config.go:41]: swap this out for a permalink plz

File: nodeadm/internal/containerd/config.go [hunk 2]
```diff
@@ -35,16 +36,28 @@ func writeContainerdConfig(cfg *api.NodeConfig) error {
...
+		if err := toml.Unmarshal(containerdConfig, &systemContainerdConfigMap); err != nil {
+			return err
+		}
+		if err := mergo.Merge(&systemContainerdConfigMap, userContainerdConfigMap, mergo.WithOverride); err != nil {
+			return err
+		}
+		containerdConfig, err = toml.Marshal(systemContainerdConfigMap)
+		if err != nil {
+			return err
+		}
```
Comment (cartermckinnon) [inline nodeadm/internal/containerd/config.go:57]: This is really similar to what's happening in `internal/api/merge.go`, can we add one utility method to `utils` or something?

File: nodeadm/internal/containerd/config.go [hunk 3]
```diff
@@ -35,16 +36,28 @@ func writeContainerdConfig(cfg *api.NodeConfig) error {
...
-	if err := util.WriteFileWithDir(containerdConfigFile, containerdConfig, containerdConfigPerm); err != nil {
-		return err
-	}
+	// because the logic in containerd's import merge decides to completely
+	// overwrite entire sections, we want to implement this merging ourselves.
+	// see: https://github.com/containerd/containerd/blob/main/cmd/containerd/server/config/config.go#L407-L431
 	if len(cfg.Spec.Containerd.Config) > 0 {
-		containerConfigImportPath := filepath.Join(containerdConfigImportDir, "00-nodeadm.toml")
-		zap.L().Info("Writing user containerd config to drop-in file..", zap.String("path", containerConfigImportPath))
-		return util.WriteFileWithDir(containerConfigImportPath, []byte(cfg.Spec.Containerd.Config), containerdConfigPerm)
```
Comment (cartermckinnon) [inline nodeadm/internal/containerd/config.go:45]: You can get rid of the `containerConfigImportPath` constant now right?

Comment (ndbaker1) [reply nodeadm/internal/containerd/config.go:45]: yup!

File: nodeadm/internal/containerd/config.go (hunk 2 above)
Comment (ndbaker1) [reply nodeadm/internal/containerd/config.go]: i reworked the current merge helper to be a lot more parameterized, take a look and let me know what you think  
I think it yields the least duplicate code for the least work 😬

**Comment by ndbaker1** [comment]
/ci
+workflow:os_distros al2023

**Review by cartermckinnon — APPROVED** [review]
LGTM

**Pull request merged**